app/models/katello/glue/candlepin/pool.rb

```diff
@@ -27,28 +27,58 @@ def self.included(base)
     end
 
     module ClassMethods
-      def candlepin_data(cp_id)
+      def candlepin_data(cp_id, rescue_gone = false)
         Katello::Resources::Candlepin::Pool.find(cp_id)
+      rescue Katello::Errors::CandlepinPoolGone => e
+        raise e unless rescue_gone
+
+        if (pool_id = ::Katello::Pool.find_by_cp_id(cp_id)&.id)
+          Katello::EventQueue.push_event(::Katello::Events::DeletePool::EVENT_TYPE, pool_id)
+          Rails.logger.warn("Sending pool delete event for missing candlepin pool cp_id=#{cp_id}")
+        end
+        {}
       end
 
       def get_for_owner(organization)
         Katello::Resources::Candlepin::Pool.get_for_owner(organization, true)
       end
 
       def import_pool(cp_pool_id, index_hosts = true)
-        json = candlepin_data(cp_pool_id)
-        ::Katello::Util::Support.active_record_retry do
-          pool = Katello::Pool.where(:cp_id => cp_pool_id, :organization => Organization.find_by(:label => json['owner']['key'])).first_or_create
-          pool.backend_data = json
-          pool.import_data(index_hosts)
+        Katello::Logging.time("import candlepin pool", data: { cp_id: cp_pool_id }) do
+          json = candlepin_data(cp_pool_id)
+
+          org = Organization.find_by(label: json['owner']['key'])
+          fail("Organization with label #{json['owner']['key']} wasn't found while importing Candlepin pool") unless org
+
+          subscription = determine_subscription(
+            product_id: json['productId'],
+            source_stack_id: json['sourceStackId'],
+            organization: org
+          )
+
+          fail("No subscription found") unless subscription
```

Comment:
what would the result be in this exception being thrown?  Would it be better to log it and continue?  (i don't actually know the answer, just curious what the implications are)

Reply:
Somewhat unnecessary but does allow to fail faster. The new validations catch the problem and import_data should raise an error when it calls `save!` if I remove this line.

But now I'm wondering now if `first_or_create` should be `first_or_create!` before going thru `import_data` at all. I think that's the way to go here given the new validations

Reply:
sounds good!